tests/e2e/e2e_test.go

```diff
@@ -111,3 +109,257 @@ func (s *IntegrationTestSuite) TestBankTokenTransfer() {
+	s.T().Logf("Proposal number: %d", proposalCounter)
+
+	s.fundCommunityPool(chainAAPIEndpoint, sender)
+
+	s.T().Logf("Submitting Legacy Gov Proposal: Community Spend Funding Gov Module")
+	s.submitLegacyProposalFundGovAccount(chainAAPIEndpoint, sender)
+	s.T().Logf("Depositing Legacy Gov Proposal: Community Spend Funding Gov Module")
+	s.depositGovProposal(chainAAPIEndpoint, sender, proposalCounter)
+	s.T().Logf("Voting Legacy Gov Proposal: Community Spend Funding Gov Module")
+	s.voteGovProposal(chainAAPIEndpoint, sender, proposalCounter, "yes", false)
+
+	initialGovBalance, err := getSpecificBalance(chainAAPIEndpoint, govModuleAddress, photonDenom)
+	s.Require().NoError(err)
+	proposalCounter++
+
+	s.T().Logf("Submitting Gov Proposal: Sending Tokens from Gov Module to Recipient")
+	s.submitNewGovProposal(chainAAPIEndpoint, sender, proposalCounter, "/root/.gaia/config/proposal_2.json")
+	s.T().Logf("Depositing Gov Proposal: Sending Tokens from Gov Module to Recipient")
+	s.depositGovProposal(chainAAPIEndpoint, sender, proposalCounter)
+	s.T().Logf("Voting Gov Proposal: Sending Tokens from Gov Module to Recipient")
+	s.voteGovProposal(chainAAPIEndpoint, sender, proposalCounter, "yes", false)
+	s.Require().Eventually(
+		func() bool {
+			sendGovAmount := sdk.NewInt64Coin(photonDenom, 10)
+			newGovBalance, err := getSpecificBalance(chainAAPIEndpoint, govModuleAddress, photonDenom)
+			s.Require().NoError(err)
+
+			recipientBalance, err := getSpecificBalance(chainAAPIEndpoint, govSendMsgRecipientAddress, photonDenom)
+			s.Require().NoError(err)
+			return newGovBalance.IsEqual(initialGovBalance.Sub(sendGovAmount)) && recipientBalance.Equal(initialGovBalance.Sub(newGovBalance))
+		},
+		15*time.Second,
+		5*time.Second,
+	)
+
+}
+
+func (s *IntegrationTestSuite) TestGovSoftwareUpgrade() {
+	chainAAPIEndpoint := fmt.Sprintf("http://%s", s.valResources[s.chainA.id][0].GetHostPort("1317/tcp"))
+	senderAddress, err := s.chainA.validators[0].keyInfo.GetAddress()
+	s.Require().NoError(err)
+	sender := senderAddress.String()
+	height := s.getLatestBlockHeight(s.chainA, 0)
+	proposalHeight := height + 35
+	proposalCounter++
+
+	s.T().Logf("Writing proposal %d on chain %s", proposalCounter, s.chainA.id)
+	s.writeGovUpgradeSoftwareProposal(s.chainA, proposalHeight)
+
+	s.T().Logf("Submitting Gov Proposal: Software Upgrade")
+	s.submitNewGovProposal(chainAAPIEndpoint, sender, proposalCounter, "/root/.gaia/config/proposal_3.json")
+	s.T().Logf("Depositing Gov Proposal: Software Upgrade")
+	s.depositGovProposal(chainAAPIEndpoint, sender, proposalCounter)
+	s.T().Logf("Weighted Voting Gov Proposal: Software Upgrade")
+	s.voteGovProposal(chainAAPIEndpoint, sender, proposalCounter, "yes=0.8,no=0.1,abstain=0.05,no_with_veto=0.05", true)
+
+	s.verifyChainHaltedAtUpgradeHeight(s.chainA, 0, proposalHeight)
+	s.T().Logf("Successfully halted chain at height %d", proposalHeight)
+
+	currentChain := s.chainA
+
+	for valIdx := range currentChain.validators {
```

Comment:
not sure if we have to delete all the validators and re-launch them, shouldn't we just be able to tell them to start running agin? Maybe there's a benefit to starting from scratch? In another scenario, if anyone were to join the network and replay from 0, they would be forced to halt at the upgrade height and restart themselves but this won't happen here right? Is there any state from earlier transactions we want to keep? Maybe it's ok to nuke em and keep the tests running, just might be confusing if anyone adds an earlier test that expects that state to persist past the governance tests.

Reply:
discussed out of band with lauren, just wanted to update here that i think it's better to keep as nuking them afterall. it preserves the thinking that the tests should in fact be able to run independent of each other. and it's a little sloppy stopping and starting the chain again with the `--skip-upgrade` flag.